#### RFM_Analysis Loan Book Dataset 

##### Columns Data;
- Account Number: Unique identifier for the loan account.
- Customer ID: Unique identifier for the customer who took the loan.
- Loan Account Status: The current state of the loan, indicating performance or closure.
- Disbursement Date: The date the loan amount was issued or disbursed to the customer.
- Principal Amount: The original amount of money lent to the borrower (before any interest or fees).
- Tenure (days): Duration of the loan in days, showing how long the borrower was expected to repay the loan.
- Days past due: Number of days the loan payment is overdue, measuring how late the repayment is.
- Aging: Loan aging bucket for categorizing delinquency or time since default. e.g W/O indicates written-off loans.
- Number of Loans Taken: Total number of loans the customer has taken with the lender.
- Loan_No_Index: The sequence of the current loan relative to the customer e.g., 1st loan, 2nd loan...
- CRB Stage: Stage classification from the Credit Reference Bureau, indicating the credit health or risk tier e.g., Stage 1, Stage 2.
- ECTO Value: Score or predictive value associated with risk, customer behavior, or expected credit loss.
- Risk Score: Numeric value assessing the riskiness of the customer or loan. Higher scores may indicate greater risk.
- CRB Worst Days: The maximum number of days a customer has ever been overdue, as reported by CRB.
- CRB Category: A qualitative classification from CRB, such as Bad Bureau, representing the borrower's historical credit behavior.

In [4]:
import pandas as pd 
from datetime import datetime 

In [5]:
file = (r'D:\dATA sCIENCE\eXPLORE pY\lOCAL dATASETS\RFM-Loan-Book-Dataset-2024-2025\data\raw\rfm_analysis loan book dataset.xlsx')
df = pd.read_excel(file)
df.head(10)

Account Number  Customer ID Loan Account Status Disbursement Date  \
0     AZ10021080    100250840           Write Off        2024-10-02   
1     AZ10010507    100250432           Write Off        2024-10-10   
2     AZ10012735    100253836           Write Off        2024-10-04   
3     AZ10020144    100253615           Write Off        2024-10-09   
4     AZ10012926    100252507             Matured        2024-10-06   
5     AZ10011319    100250860             Matured        2024-10-12   
6     AZ10020677    100250835           Write Off        2024-10-04   
7     AZ10015932    100250531           Write Off        2024-10-08   
8     AZ10010982    100250542           Write Off        2024-10-03   
9     AZ10012948    100251331           Write Off        2024-10-10   

   Principal Amount  Tenure (days)  Days past due Aging  \
0      6.285126e+05              8             91   W/O   
1      3.663830e+06              6             91   W/O   
2      5.140163e+04             17             91   W/O   
3      2.085975e+06             12             91   W/O   
4      7.460801e+05             13              0     0   
5      6.621669e+04             14              0     0   
6      4.086957e+06             15             91   W/O   
7      8.942859e+05             12             91   W/O   
8      1.812559e+07             14             91   W/O   
9      7.754575e+05             17             91   W/O   

   Number of Loans Taken  Loan_No_Index CRB Stage    ECTO Value  Risk Score  \
0                      4              4   Stage 2           NaN         NaN   
1                      3              3   Stage 2           NaN         NaN   
2                      1              1   Stage 2           NaN         NaN   
3                      1              1   Stage 2           NaN         NaN   
4                      3              1   Stage 2  1.411374e+07        17.0   
5                      1              1   Stage 2           NaN         NaN   
6                      1              1   Stage 2           NaN         NaN   
7                      1              1   Stage 2           NaN         NaN   
8                      1              1   Stage 2           NaN         NaN   
9                      2              2   Stage 2           NaN         NaN   

  CRB Worst Days CRB Category  
0             60   Bad Bureau  
1             88   Bad Bureau  
2             84   Bad Bureau  
3             90   Bad Bureau  
4             53   Bad Bureau  
5             53   Bad Bureau  
6             53   Bad Bureau  
7             84   Bad Bureau  
8             84   Bad Bureau  
9             84   Bad Bureau

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7228 entries, 0 to 7227
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   Account Number         7228 non-null   object        
 1   Customer ID            7228 non-null   int64         
 2   Loan Account Status    7228 non-null   object        
 3   Disbursement Date      7228 non-null   datetime64[ns]
 4   Principal Amount       7228 non-null   float64       
 5   Tenure (days)          7228 non-null   int64         
 6   Days past due          7228 non-null   int64         
 7   Aging                  7228 non-null   object        
 8   Number of Loans Taken  7228 non-null   int64         
 9   Loan_No_Index          7228 non-null   int64         
 10  CRB Stage              6494 non-null   object        
 11  ECTO Value             3187 non-null   float64       
 12  Risk Score             3187 non-null   float64       
 13  CRB

In [7]:
df.describe()

Customer ID              Disbursement Date  Principal Amount  \
count  7.228000e+03                           7228      7.228000e+03   
mean   1.002520e+08  2025-01-14 16:01:15.705589504      8.125035e+06   
min    1.002500e+08            2024-10-02 00:00:00      2.157700e+02   
25%    1.002510e+08            2024-11-28 00:00:00      8.790240e+05   
50%    1.002520e+08            2025-01-07 00:00:00      3.269230e+06   
75%    1.002529e+08            2025-03-09 00:00:00      1.027868e+07   
max    1.002539e+08            2025-05-10 00:00:00      4.996892e+07   
std    1.116629e+03                            NaN      1.113536e+07   

       Tenure (days)  Days past due  Number of Loans Taken  Loan_No_Index  \
count    7228.000000    7228.000000            7228.000000    7228.000000   
mean       30.829552       7.675706               5.392225       3.088406   
min         2.000000       0.000000               1.000000       1.000000   
25%        23.000000       0.000000               3.000000       1.000000   
50%        34.000000       0.000000               5.000000       2.000000   
75%        38.000000       0.000000               7.000000       4.000000   
max        41.000000      91.000000              35.000000      35.000000   
std         9.674029      24.117620               3.984734       2.737248   

         ECTO Value   Risk Score  
count  3.187000e+03  3187.000000  
mean   3.551973e+08    18.216505  
min    0.000000e+00    12.000000  
25%    2.083698e+07    16.000000  
50%    8.011665e+07    18.000000  
75%    2.701901e+08    19.000000  
max    3.391474e+10    97.000000  
std    1.472168e+09     4.194750

##### 1]Recency(R)Showing how recently a customer was active:
- Represented by 'Disbursement Date' - to calculate days since the last loan was disbursed to each customer.
##### 2]Frequency(F)Showing how often a customer takes loans:
- Represented by 'Number of Loans Taken' - to count the number of loans per customer from the data.
##### 3]Monetary(M)Showing how much value the customer brings:
- Represented by the total principal amount disbursed to a customer.


In [8]:
latest_df = df['Disbursement Date'].max() # Get the latest disbursement date in the data

recency_df = df.groupby('Customer ID')['Disbursement Date'].max().reset_index() # Calculate the most recent loan per customer
recency_df['Recency'] = (latest_df - recency_df['Disbursement Date']).dt.days # Compute recency in days

frequency_df = df.groupby('Customer ID')['Account Number'].nunique().reset_index(name='Frequency') # 

monetary_df = df.groupby('Customer ID')['Principal Amount'].sum().reset_index(name='Monetary')

rfm = recency_df.merge(frequency_df, on='Customer ID').merge(monetary_df, on='Customer ID')

rfm = rfm[['Customer ID', 'Recency', 'Frequency', 'Monetary']]


##### Next Steps:
- Segment customers using quantiles or clusters.
- Score them on a scale (e.g., 1 to 5) per RFM variable.
- Use it for credit behavior profiling or marketing/retention strategies.

In [9]:
rfm['R_Score'] = pd.qcut(rfm['Recency'], 5, labels=[5, 4, 3, 2, 1]).astype(int)
rfm['F_Score'] = pd.qcut(rfm['Frequency'].rank(method='first'), 5, labels=[1, 2, 3, 4, 5]).astype(int)
rfm['M_Score'] = pd.qcut(rfm['Monetary'].rank(method='first'), 5, labels=[1, 2, 3, 4, 5]).astype(int)

rfm['RFM_Segment'] = rfm['R_Score'].astype(str) + rfm['F_Score'].astype(str) + rfm['M_Score'].astype(str)
rfm['RFM_Score'] = rfm[['R_Score', 'F_Score', 'M_Score']].sum(axis=1)

rfm.head(10)


Customer ID  Recency  Frequency      Monetary  R_Score  F_Score  M_Score  \
0    100250001       31          7  2.867315e+07        5        5        4   
1    100250003      116          2  2.043290e+07        2        3        4   
2    100250004      186          3  4.146474e+05        1        4        1   
3    100250005      210          1  2.742336e+05        1        1        1   
4    100250006      125          1  1.348571e+05        2        1        1   
5    100250008      210          1  2.003017e+05        1        1        1   
6    100250010       46          5  6.992871e+07        4        5        5   
7    100250013        1          1  5.311267e+06        5        1        3   
8    100250014      218          1  1.584446e+06        1        1        2   
9    100250015      142          3  6.434034e+07        2        4        5   

  RFM_Segment  RFM_Score  
0         554         14  
1         234          9  
2         141          6  
3         111          3  
4         211          4  
5         111          3  
6         455         14  
7         513          9  
8         112          4  
9         245         11

##### Interpretation:
R_Score: 5 = very recent activity; 1 = old/inactive
F_Score: 5 = very frequent borrower; 1 = rarely borrows
M_Score: 5 = large total loan amounts; 1 = low loan amounts

RFM_Score: Total of all 3 scores (max = 15)
RFM_Segment: A string useful for segmentation, e.g. "554" means: Recent, Frequent, High Value

In [10]:
def customer_segment(rfm):
    if rfm['RFM_Score'] >= 13:
        return 'Champions'
    elif rfm['RFM_Score'] >= 10:
        return 'Loyal Customers'
    elif rfm['RFM_Score'] >= 7:
        return 'Potential Loyalist'
    elif rfm['RFM_Score'] >= 5:
        return 'At Risk'
    else:
        return 'Hibernating'
    

rfm['Segment'] = rfm.apply(customer_segment, axis=1)

rfm[['Customer ID', 'RFM_Segment', 'RFM_Score', 'Segment']].head(10)

Customer ID RFM_Segment  RFM_Score             Segment
0    100250001         554         14           Champions
1    100250003         234          9  Potential Loyalist
2    100250004         141          6             At Risk
3    100250005         111          3         Hibernating
4    100250006         211          4         Hibernating
5    100250008         111          3         Hibernating
6    100250010         455         14           Champions
7    100250013         513          9  Potential Loyalist
8    100250014         112          4         Hibernating
9    100250015         245         11     Loyal Customers

##### Segment Descriptions:
Champions: Recent, frequent, and high-value borrowers — best customers.

Loyal Customers: Borrow often and bring good value.

Potential Loyalist: Could become loyal with some engagement.

At Risk: Value is average, and recency is worsening.

Hibernating: Old, inactive, or low-value customers.

In [11]:
Segment_Counts  = rfm['Segment'].value_counts().reset_index()
Segment_Counts.columns = ['Segment', 'Customer Count']
Segment_Counts

Segment  Customer Count
0  Potential Loyalist             871
1     Loyal Customers             738
2           Champions             586
3             At Risk             581
4         Hibernating             256

##### Insights:

Potential Loyalists make up the largest group — strong candidates for conversion to loyal customers.

Champions and Loyal Customers together are over 1,300 high-value clients worth maintaining.

Hibernating and At Risk groups might need reactivation or recovery strategies.

In [13]:
rfm.to_csv(r'D:\dATA sCIENCE\eXPLORE pY\lOCAL dATASETS\RFM-Loan-Book-Dataset-2024-2025\data\processed\rfm.csv')

In [ ]:
https://github.com/OkotPascal?tab=repositories